### ISSUE TITLE
Ctrl+C handling.

### ISSUE THREAD

**Issue by kajojify** [ronf/asyncssh#103]
Asyncssh 1.10.1
Python 3.5.2
OS: Ubuntu 16.04 LTS

Hello! Sorry for disturbing you. 
I am not sure, that it is asyncssh bug, so sorry if it is something from my side. 
It seems, asyncssh can't handle `KeyboardInterrupt` exception appropriately.

## How to reproduce: 
I took [this example](https://asyncssh.readthedocs.io/en/latest/#simple-server) from asyncssh documentation and modified it a little bit:
```
import asyncio, asyncssh, sys


class MySSHServer(asyncssh.SSHServer):
    def connection_made(self, conn):
        print('SSH connection received from %s.' %
                  conn.get_extra_info('peername')[0])

    def connection_lost(self, exc):
        if exc:
            print('SSH connection error: ' + str(exc), file=sys.stderr)
        else:
            print('SSH connection closed.')

    def begin_auth(self, username):
        return True

    def password_auth_supported(self):
        return True

    def validate_password(self, username, password):
        print(username, password)
        return False

async def start_server():
    await asyncssh.create_server(MySSHServer, '', 22,
                                 server_host_keys=['ssh.key'])

loop = asyncio.get_event_loop()

try:
    loop.run_until_complete(start_server())
except (OSError, asyncssh.Error) as exc:
    sys.exit('Error starting server: ' + str(exc))

loop.run_forever()
```
I started this script using: `sudo python3 cd.py`
Then I tried to "bruteforce" it using Hydra v8.6 (version doesn't matter I think):
`hydra -l ssh_user -P passlist.txt ssh://localhost`
During the "bruteforcing" I pressed `Ctrl+C` in the terminal, where ssh.py was running. I got this:
```
ssh_user 15935
ssh_user 15031978
ssh_user 15101993
ssh_user 130989
ssh_user 15051984
ssh_user 15061981
ssh_user 14061979
^CTraceback (most recent call last):
  File "cd.py", line 36, in <module>
    loop.run_forever()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 345, in run_forever
    self._run_once()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 1312, in _run_once
    handle._run()
  File "/usr/lib/python3.5/asyncio/events.py", line 125, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 374, in _run_task
    yield from coro
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/auth.py", line 799, in _start
    password = saslprep(password.decode('utf-8'))
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/saslprep.py", line 106, in saslprep
    return _stringprep(s, True, _map_saslprep, 'NFKC', prohibited, True)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/saslprep.py", line 62, in _stringprep
    if stringprep.in_table_a1(c):
  File "/usr/lib/python3.5/stringprep.py", line 13, in in_table_a1
    if unicodedata.category(code) != 'Cn': return False
KeyboardInterrupt
Task was destroyed but it is pending!
task: <Task pending coro=<SSHConnection._run_task() running at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366>>
Task was destroyed but it is pending!
task: <Task pending coro=<SSHConnection._run_task() running at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366>>
Task was destroyed but it is pending!
task: <Task pending coro=<SSHConnection._run_task() running at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366>>
Task was destroyed but it is pending!
task: <Task pending coro=<SSHConnection._run_task() running at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366>>
Task exception was never retrieved
future: <Task finished coro=<SSHConnection._run_task() done, defined at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366> exception=KeyboardInterrupt()>
Traceback (most recent call last):
  File "cd.py", line 36, in <module>
    loop.run_forever()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 345, in run_forever
    self._run_once()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 1312, in _run_once
    handle._run()
  File "/usr/lib/python3.5/asyncio/events.py", line 125, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 374, in _run_task
    yield from coro
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/auth.py", line 799, in _start
    password = saslprep(password.decode('utf-8'))
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/saslprep.py", line 106, in saslprep
    return _stringprep(s, True, _map_saslprep, 'NFKC', prohibited, True)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/saslprep.py", line 62, in _stringprep
    if stringprep.in_table_a1(c):
  File "/usr/lib/python3.5/stringprep.py", line 13, in in_table_a1
    if unicodedata.category(code) != 'Cn': return False
KeyboardInterrupt
Task was destroyed but it is pending!
task: <Task pending coro=<SSHConnection._run_task() running at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366>>
Task was destroyed but it is pending!
task: <Task pending coro=<SSHConnection._run_task() running at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366>>
```
According to [official asyncio documentation](https://docs.python.org/3/library/asyncio-dev.html#pending-task-destroyed) it can be considered as a bug.

Trying to use try-except at the end of the script:
```
try:
    loop.run_forever()
except KeyboardInterrupt:
    pass
```
I get this: 
```
ssh_user zeyne
ssh_user zeus69
ssh_user ZEUS
^CTask was destroyed but it is pending!
task: <Task pending coro=<SSHConnection._run_task() running at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366>>
Task exception was never retrieved
future: <Task finished coro=<SSHConnection._run_task() done, defined at /usr/local/lib/python3.5/dist-packages/asyncssh/connection.py:366> exception=KeyboardInterrupt()>
Traceback (most recent call last):
  File "cd.py", line 36, in <module>
    loop.run_forever()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 345, in run_forever
    self._run_once()
  File "/usr/lib/python3.5/asyncio/base_events.py", line 1312, in _run_once
    handle._run()
  File "/usr/lib/python3.5/asyncio/events.py", line 125, in _run
    self._callback(*self._args)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 239, in _step
    result = coro.send(None)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 374, in _run_task
    yield from coro
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/auth.py", line 818, in _start
    self.send_failure()
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/auth.py", line 467, in send_failure
    self._conn.send_userauth_failure(partial_success)
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/connection.py", line 1004, in send_userauth_failure
    NameList(get_server_auth_methods(self)),
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/packet.py", line 65, in NameList
    return String(b','.join(value))
  File "/usr/local/lib/python3.5/dist-packages/asyncssh/packet.py", line 49, in String
    return len(value).to_bytes(4, 'big') + value
KeyboardInterrupt
```
Thank you for your attention! 
Have a nice day!

**Comment by ronf** [comment]
Since asyncssh is a library, it's up to the calling program to clean up any SSH connections it has opened and wait for them to close before exiting the event loop when catching an exception like KeyboardInterrupt or otherwise deciding to exit while connections are still active.

Generally, if you really want to exit abruptly, you'll want to keep track of all of the connections you've opened and call abort() on them to get them to terminate as quickly as possible, without waiting for any buffered data to be delivered. Then, after doing that, you'd want to await on wait_closed() for each of these connections (or all of them in parallel, using something like asyncio.gather()). When that's complete, it should be safe to exit the event loop and end the process and you should no longer see the errors about tasks being destroyed while still pending.

**Comment by ronf** [comment]
Closing this issue for now - please re-open it or open a new issue if you need more than the explanation above.
